cupyx/scipy/sparse/linalg/solve.py

```diff
@@ -65,3 +66,266 @@ def lsqr(A, b):
     x = x.astype(numpy.float64)
     ret = (x, None, None, None, None, None, None, None, None, None)
     return ret
+
+
+def bicgstab(A, b, x0=None, M=None, tol=1e-5, maxiter=numpy.inf,
+             callback=None, atol='legacy'):
+    """Solves linear system using an iterative solver BiCGSTAB.
+
+    See https://docs.nvidia.com/cuda/cusparse/#csrilu02_solve and
+    https://docs.nvidia.com/cuda/incomplete-lu-cholesky/index.html.
+
+    Find the solution to a large, sparse, linear system of equations.
+    The function approximates ``Ax = b`` using BiCGSTAB iterative method.
+
+    Args:
+        A (cupy.ndarray or cupyx.scipy.sparse.csr_matrix): The input matrix
+            with dimension ``(N, N)``
+        b (cupy.ndarray): Right-hand side vector.
+        x0 (cupy.ndarray): Initial guess (defaults to zeros)
+        M (tuple(callable, optional(info))): GPU preconditioner function
+            and info object for preconditioner (to destroy after running)
+        tol (float): Tolerance for convergence, exits when
+                     residual norm(r) < tol * norm(b)
+        maxiter (int): maximum number of iterations (defaults to `numpy.inf`)
+        callback (callable): a callback function to call at each iteration
+                            (unlike scipy, it accepts iteration and
+                            residual as well as current vector `x`)
```

Comment:
The rationale is that getting residual information may be useful for debugging and it might take longer since most of the iteration when unpreconditioned is probably the sparse matvecs which we also need to get the residual... But I'm fine sticking with `scipy`'s convention if you like!